### PR TITLE
sql: represent operators as text in AST

### PR DESCRIPTION
**Pull request by benesch**
Rather than representing operators via two enums, UnaryOperator and
BinaryOperator, represent them as a string that stores the literal
operator symbol. This brings our implementation more in line with
PostgreSQL, which permits users to create arbitrary operators out of a
set of allowable operator symbols.

Also rework the function selection API a bit to share more code.
Operators and functions are now basically indistinguishable from the
perspective of the sql::plan::func module.

The goal here is to make room for PostgreSQL's qualified operator
syntax:

    SELECT 1 OPERATOR(pg_catalog.+) 2

though actually implementing this syntax is left to a forthcoming
commit.

---
This change is [Reviewable](https://reviewable.io/reviews/materializeinc/materialize/4653)

### CODE REVIEW

**Review by JLDLaughlin — APPROVED** [review]
looks good to me!

File: src/sql/src/plan/func.rs [hunk 1]
```diff
@@ -1547,33 +1544,47 @@ pub fn resolve(scx: &StatementContext, name: &PartialName) -> Result<&'static Fu
...
-    let mut cexprs = Vec::new();
-    for arg in args {
-        let cexpr = query::plan_expr(ecx, arg)?;
-        cexprs.push(cexpr);
-    }
-
-    let ident = &name.to_string();
-    ArgImplementationMatcher::select_implementation(ident, func_err_string, ecx, impls, cexprs)
+    let types: Vec<_> = args.iter().map(|e| ecx.scalar_type(e)).collect();
+    ArgImplementationMatcher::select_implementation(ecx, spec, impls, args, &types).with_context(
```
Comment (sploiselle) [inline src/sql/src/plan/func.rs:1555]: TIL about `with_context`; this is great.

**Review by sploiselle — APPROVED** [review]
LGTM

**Comment by benesch** [comment]
🎉 thanks for the reviews!

**Pull request merged**